doc/designs/quic-design/synthesis.md

```diff
@@ -0,0 +1,507 @@
+
+/*
+ * (There will of course be BIO_met_get/set functions corresponding to all of
+ * these, which are not shown here)
+ */
+
+/*
+ * These functions are used to setup local address support for a BIO_dgram.
+ *
+ * BIO_dgram_get_local_addr_cap determines if a BIO can support local address
+ * support (i.e., the local field of BIO_MSG being non-NULL). It returns 1 if
+ * support is available.
+ */
+int BIO_dgram_get_local_addr_cap(BIO *b);
+/* BIO_CTRL_DGRAM_GET_LOCAL_ADDR_CAP */
+
+/*
+ * BIO_dgram_get/set_local_addr_enable turns local address support on and off.
+ * The reason we force this to be done up-front is because it requires a
+ * setsockopt, and we do not want to do this lazily because this will introduce
+ * the need for locking into BIO_sendmmsg and BIO_recvmmsg.
+ *
+ * BIO_dgram_set_local_addr_enable() returns 1 on success and 0 otherwise (e.g.
+ * if support is not available).
+ *
+ * BIO_dgram_get_local_addr_enable() writes 1 to *enable if support is
+ * currently enabled and 0 otherwise and returns 1.
+ */
+int BIO_dgram_set_local_addr_enable(BIO *b, int enable);
+// BIO_CTRL_DGRAM_SET_LOCAL_ADDR_ENABLiE
+
+int BIO_dgram_get_local_addr_enable(BIO *b, int *enable);
+// BIO_CTRL_DGRAM_GET_LOCAL_ADDR_ENABLE
+```
+
+Post-MVP Functionality
+----------------------
+
+```c
+/* ========================================================
+ * Alternate API to avoid copying. Not for MVP.
+ */
+int BIO_get_dgram(BIO *bio, void **buf, size_t *buf_len);
+int BIO_return_dgram(BIO *bio, void *buf, size_t buf_len);
+int BIO_put_dgram(BIO *bio, const void *buf, size_t buf_len,
+                  void (*free_cb)(void *buf, size_t buf_len, void *arg),
+                  void *arg);
+
+/* ========================================================
+ * API for multiple streams. Not for MVP.
+ */
+
+/*
+ * Create a new SSL object representing a single additional stream.
+ *
+ * There is no need to call SSL_connect on the resulting object, and
+ * any such call is a no-op.
+ *
+ * For QUIC:
+ *   Creates a new stream. Can be used on client or server. If the SSL_FLAG_UNI
+ *   flag is set, the created stream is unidirectional, otherwise it is
+ *   bidirectional.
+ *
+ * For TLS and DTLS clients:
+ *   Attempts a resumption handshake. The flags argument is ignored.
+ *
+ * For TLS and DTLS servers:
+ *   Always fails.
+ */
+#define SSL_FLAG_UNI    1
+SSL *SSL_new_stream(SSL *ssl, uint64_t flags);
+
+/*
+ * Create a new SSL object representing an additional stream which has created
+ * by the peer.
+ *
+ * There is no need to call SSL_accept on the resulting object, and
+ * any such call is a no-op.
+ *
+ * For QUIC:
+ *   Checks if a new stream has been created by the peer. If it has, creates a
+ *   new SSL object to represent it and returns it. Otherwise, returns NULL.
+ *
+ * For all other methods:
+ *   Returns NULL.
+ *
+ * The flags argument is unused and should be set to zero.
+ *
+ * TODO: Do we want a WAIT flag for synchronous blocking here?
```

Comment:
IMO not necessary for first iteration.

However it would also be nice to have a possibility to set a callback to be called to notify the application that there is a stream to be accepted.